refactor(common): update AsyncPipe to support PromiseLike

tjlav5 opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/main/contributing-docs/commit-message-guidelines.md
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.dev application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
AsyncPipe today works with Observable|Subscribale|Promise.

Issue Number: N/A


## What is the new behavior?
Widen Promise support with duck-typing such that third-party libraries that may not be using native Promises can still be passed into the AsyncPipe.


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

tjlav5 committed:
refactor(common): update AsyncPipe to support PromiseLike

Updates the AsyncPipe to work with PromiseLike, in addition to Promise.

crisbeto reviewed (approved):
Reviewed-for: public-api
Reviewed-for: fw-general

pkozlowski-opensource reviewed (approved):
LGTM

Reviewed-for: public-api 
Reviewed-for: fw-general

JeanMeche commented:
There are legit breakages in G3, some projects extend the `AsyncPipe`.

tjlav5 committed:
refactor(common): update AsyncPipe to explicitly accept Promise | PromiseLike

There is a chance that external users have "extended" AsyncPipe and are expecting to
pass in a Promise instance.

tjlav5 committed:
refactor(common): update AsyncPipe to explicitly accept Promise | PromiseLike

There is a chance that external users have "extended" AsyncPipe and are expecting to
pass in a Promise instance.

tjlav5 commented:
ack - and considering that external users may be extending AsyncPipe in a similar manner, I suppose it's safe to just expand the transform type to be `Promise | PromiseLike`

JeanMeche commented:
@tjlav5 fwiw, extending is not part of our public API policy. 

https://github.com/angular/angular/blob/6d3849ffc85cc52b0eba3836b24f10895621acbe/contributing-docs/public-api-surface.md?plain=1#L55-L59

Your alternative seems interesting though !

kirjs reviewed (approved):
Reviewed-for: public-api
Reviewed-for: fw-general

kirjs reviewed (commented):
Awesome, thanks for the contribution, do we need to update the docs for this?

tjlav5 commented:
I'm not sure... the signatures changed, is that good enough? If the docs still mention promises, I think it's technically still accurate.

JeanMeche commented:
I think we're good with the API docs that get automatically updated.

tjlav5 committed:
refactor(common): update AsyncPipe to explicitly accept PromiseLike only

With the fixed in g3, we can keep the type limited to PromiseLike
instead of Promise|PromiseLike

tjlav5 commented:
FYI - I believe I've fixed all of the failing g3 code, and reverted this PR back to the original proposal of Promise -> PromiseLike

JeanMeche commented:
Could you link a TGP that confirms that ? (so that the caretaker can safely merge the PR and submit the CL).

tjlav5 commented:
Is the presubmit linked here sufficient: http://cl/741629375?

JeanMeche commented:
Thanks it's a good start, I can tree trigger TGP over the week to be safe. 👍

JeanMeche commented:
[TGP](https://fusion2.corp.google.com/presubmit/741168330/OCL:741168330:BASE:742489369:1743474412713:d4b94011) looks fine

thePunderWoman commented:
This PR was merged into the repository by commit b927ce8ec9cf1de1f2d79c817695f1b5198b3451.

The changes were merged into the following branches: main
